kv/memberlist/tcp_transport.go

```diff
@@ -424,27 +446,35 @@ func (t *TCPTransport) getAdvertisedAddr() string {
 // WriteTo is a packet-oriented interface that fires off the given
 // payload to the given address.
 func (t *TCPTransport) WriteTo(b []byte, addr string) (time.Time, error) {
-	t.sentPackets.Inc()
-	t.sentPacketsBytes.Add(float64(len(b)))
+	if t.shutdown.Load() == 1 {
+		return time.Time{}, errors.New("transport is shutting down")
+	}
```

Comment:
I think there is a race between WriteTo and Shutdown. `t.shutdown.Load()` may return 0, but by the time `WriteTo` gets to writing to channel, it may be closed.

We should check `shutdown` flag and write to the channel inside the same lock. (It would mean changing the `shutdown` flag into simple bool and protect it with a mutex).

In `Shutdown` method, we should take a lock, set shutdown to true, close the channel, release the lock, and then wait for all goroutines to stop. (No need to hold lock for that anymore)